fdbclient/BlobCipher.cpp

```diff
@@ -1003,6 +1047,55 @@ StringRef EncryptBlobCipherAes265Ctr::encrypt(const uint8_t* plaintext,
 	return encryptBuf.substr(0, plaintextLen);
 }
 
+void EncryptBlobCipherAes265Ctr::encryptInplace(uint8_t* plaintext,
+                                                const int plaintextLen,
+                                                BlobCipherEncryptHeaderRef* headerRef) {
+	double startTime = 0.0;
+	if (CLIENT_KNOBS->ENABLE_ENCRYPTION_CPU_TIME_LOGGING) {
+		startTime = timer_monotonic();
+	}
+
+	int bytes{ 0 };
+	if (EVP_EncryptUpdate(ctx, plaintext, &bytes, plaintext, plaintextLen) != 1) {
+		TraceEvent(SevWarn, "BlobCipherEncryptUpdateFailed")
+		    .detail("BaseCipherId", textCipherKey->getBaseCipherId())
+		    .detail("EncryptDomainId", textCipherKey->getDomainId());
+		throw encrypt_ops_error();
+	}
+
+	// Padding should be 0 for AES CTR mode, so encryptUpdate() should encrypt all the data
+	if (bytes != plaintextLen) {
```

Comment:
this would work _if_ the memory buffer is aligned to AES_BLOCK_SIZE; is not then not sure if the results are acceptable? 
If above hold true, it would be good to force such an invariant in the function and maybe ASSERT on it?

Also, this requirement should be advertised as the part of the API for the clients to be aware of.
Further, we should rename the method to `encryptInPlaceAigned` or something similar to make it explicit. 

Thoughts?

Reply:
@sfc-gh-yiwu had raised the same question and we did some investigation on that: From the tests we did and double check with the [document](https://www.openssl.org/docs/man3.1/man3/EVP_CIPHER_get_block_size.html), for AES CTR mode, the `block_size` is always 1, which doesn't require padding nor memory alignment. (We also verified that the padding is required if we change CTR mode to CBC [`EVP_aes_256_cbc() here`. Then we need to allocate extra space `AES_BLOCK_SIZE` for padding.](https://github.com/apple/foundationdb/blob/eac757d1861f80a767c15da3fd0755a2a8b7851a/fdbclient/BlobCipher.cpp#L856)).
more document for that:
https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation
https://www.highgo.ca/2019/08/08/the-difference-in-five-modes-in-the-aes-encryption-algorithm/
And here is one of the tests:
https://github.com/sfc-gh-jazhuang/foundationdb/commit/a2b7a9dd6fea148a9fec1f7cabb48d8a21039594#diff-74240a17fd3787bf5b7b99522cd19290378f76c7f14efaf29be9ab8699dbe92c